File: prev-2024/2966.divide-array-into-arrays-with-max-difference.js
/*
 * @lc app=leetcode id=2966 lang=javascript
 *
 * [2966] Divide Array Into Arrays With Max Difference
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[][]}
 */
var divideArray = function (nums, k) {
  nums.sort((a, b) => a - b);

  /** @type {[number, number, number][]} */
  const result = [];

  for (let i = 0; i < nums.length - 2; i += 3) {
    const min = nums[i];
    const max = nums[i + 2];
    if (max - min <= k) {
      result.push([nums[i], nums[i + 1], nums[i + 2]]);
    } else return [];
  }

  return result;
};
// @lc code=end

File: src/2025/2966.divide-array-into-arrays-with-max-difference.cpp
/*
 * @lc app=leetcode id=2966 lang=cpp
 *
 * [2966] Divide Array Into Arrays With Max Difference
 */

#include <algorithm>
#include <vector>

// @lc code=start
class Solution {
 public:
  std::vector<std::vector<int>> divideArray(std::vector<int>& nums,
                                            const int k) {
    std::ranges::sort(nums);
    std::vector<std::vector<int>> groups;
    for (int i = 0; i < nums.size(); i += 3) {
      if (nums[i + 2] - nums[i] <= k) {
        groups.push_back({nums[i], nums[i + 1], nums[i + 2]});
      } else {
        return {};
      }
    }

    return groups;
  }
};
// @lc code=end
